lib/active_merchant/billing/gateways/worldpay.rb

```diff
@@ -1031,32 +1035,37 @@ def handle_response(response)
         end
       end
 
-      def success_from(action, raw, success_criteria)
-        success_criteria_success?(raw, success_criteria) || action_success?(action, raw)
+      def success_from(action, raw, success_or_failure_criteria)
+        if action == 'direct_inquiry'
+          return if raw[:error]
+
+          # for direct_inquiry we are looking for failed last_event
+          !success_or_failure_criteria.include?(raw[:last_event])
+        else
+          success_criteria_success?(raw, success_or_failure_criteria) || action_success?(action, raw)
+        end
       end
 
-      def message_from(success, raw, success_criteria, action)
+      def message_from(success, raw, success_or_failure_criteria, action)
         return 'SUCCESS' if success
 
-        raw[:iso8583_return_code_description] || raw[:error] || required_status_message(raw, success_criteria, action) || raw[:issuer_response_description]
+        raw[:iso8583_return_code_description] || raw[:error] || required_status_message(raw, success_or_failure_criteria, action) || raw[:issuer_response_description]
       end
 
-      # success_criteria can be:
+      # success_or_failure_criteria can be:
       #   - a string or an array of strings (if one of many responses)
       #   - An array of strings if one of many responses could be considered a
       #     success.
-      def success_criteria_success?(raw, success_criteria)
+      def success_criteria_success?(raw, success_or_failure_criteria)
         return if raw[:error]
 
-        raw[:ok].present? || (success_criteria.include?(raw[:last_event]) if raw[:last_event])
+        raw[:ok].present? || (success_or_failure_criteria.include?(raw[:last_event]) if raw[:last_event])
```

Comment:
as `nil.to_s` is equal to `""` I suggest modify this line as:
```suggestion
        raw[:ok].present? || success_or_failure_criteria.include?(raw[:last_event].to_s)
```

Reply:
I wish I saw this earlier but I'm going to leave as is for now. Since I'm deploying soon